keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,233 @@
+* Create patterns, libraries & tooling so that addons are of high quality,
+  consistent in their API surface (common fields on CRDs, use of Application
+  CRD, consistent labeling of created resources), yet are easy to build.
+* Build addons for the basic set of components, acting as a quality reference
+  implementation suitable for production use.  We aim also to demonstrate the
+  utility and explore any challenges, and to verify that the tooling does make
+  addon-development easy.
+
+
+### Non-Goals
+
+* We do not intend to mandate that all installation tools use addon operators;
+  installation tools are free to choose their own path.
+* Management of non-addons is out of scope (for example installation of end-user
+  applications, or of packaged software that is not an addon)
+
+
+## Proposal
+
+This is the current plan of action; it is based on experience gathered and work
+done for Google’s GKE-on-prem product.  However we don’t expect this will
+necessarily be directly applicable in the OSS world and we are open to change as
+we discover new requirements.
+
+* Extend kubebuilder & controller-runtime to make it easy to build operators for
+  addons
+* Build addons for the primary addons currently in the cluster/ directory, at
+  least including those required to bring up a conformant cluster.  Proposed
+  list: CoreDNS, kube-proxy, dashboard, metrics-server, localdns-agent.
+* Plug in those addons operators into kube-up / cluster-api / kubeadm / kops /
+  others (subject to those projects being interested)
+* Develop at least one addon operator outside of kubernetes/kubernetes
+  (LocalDNS-Cache?) and figure out how it can be used despite being out-of-tree
+* Investigate use of webhooks to prevent accidental mutation of child objects
+* Investigate the RBAC story for addons - currently the operator must itself
+  have all the permissions that the addon needs, which is not really
+  least-privilege.  But it is not clear how to side-step this, nor that any of
+  the alternatives would be better or more secure.
+* Investigate use of patching mechanisms (as seen in `kubectl patch` and
+  `kustomize`) to support advanced tailoring of addons.  The goal here is to
+  make sure that everyone can use the addon operators, even if they “love it but
+  just need to change one thing”.  This ensures that the addon operators
+  themselves can remain bounded in scope and complexity.  Patching will fail if
+  the underlying addon changes dramatically, so we'll likely have a "patch
+  incompatible with new version" error - and generally addons should avoid
+  gratuitously changing their structure.
+* We should develop a convention (labels / owner-refs) so that we are able to
+  discover which CRDs are cluster-addons, and there is no confusion with
+  application operators.
+
+We expect the following functionality to be common to all operators for addons:
+
+* A CRD per addon
+* Common fields in spec that define the version and/or channel
```

Comment:
What is a channel?  Maintained and supported by the community?

Reply:
A channel is what we're calling a stream of versions.  Better names welcome, but I think channel is a good name that wise people wearing red hats use...

I think it would be up to a project e.g. CoreDNS whether to support a channel for their operator and how they would define it.  But again, suggestions & recommendations are welcome.